Post about dynamic array cleanup bug (#98)
ethereum/solidity-blog: _posts/2020-10-06-solidity-dynamic-array-cleanup-bug.md

---
layout: post
published: true
title:  "Solidity Dynamic Array Cleanup Bug"

> date:   2020-04-06

Reviewer: I don't think it will happen today :)

Reviewer 2: Funnily that date is not even today, but half a year ago :-D.

Author: Ah, copy paste issues.

author: Solidity Team
category: Security Alerts
---

> On Sept 17, 2020 a bug in the Solidity code generator was found. The bug is fixed with version 0.7.3

Reviewer: Two different date formats. I think in this text, the first but with full month name would be best.

released on October 6, 2020. The bug is present in all prior versions of Solidity.

> We assigned a severity level of "medium" because the bug is easy to detect while testing the contract.

Reviewer: I would argue that this bug is one that is not trivially detectable through testing and that's why it got medium (instead of low). It is not a bug in the optimizer, so it is easier to trigger. The reasons why I would not set it to "high" are:
 - it has not been reported for several years
 - people maybe don't even rely on this behaviour and assume new array entries might be "dirty" anyway
 - increasing the length of an array without filling it with something is unlikely

Maybe we should state the severity without giving a reason for it?

elements `m` and `l` have to be cleaned by zeroing them out. However, this cleaning was not
performed properly. Specifically, after the slot corresponding to `m`, only the first packed value
was cleaned up. If this array gets resized to a length larger than `m`, the indices corresponding to
the unclean parts of the slot contained the original value, instead of 0. The resizing here is
performed by assigning to the array `length`, by a `push()` or via inline assembly. 

An example of the bug is the following:

```solidity
// SPDX-License-Identifier: GPL-3.0
// The bug is present in all versions of solidity prior to 0.7.3
pragma solidity >=0.6.0 <0.7.3;
contract C {
    uint128[] x;
    function f() public {
        x.push(42); x.push(42); x.push(42); x.push(42);
        uint128[] memory y = new uint128[](1);
        y[0] = 23;
        // This will shrink the array x to one element.
        x = y;
        // Resizing the array to length 4.
        x.push(); x.push(); x.push();
        // After resizing the array, its contents are [23, 0, 0, 42], instead of
        // [23, 0, 0, 0]. Resizing can be also be done by assigning to `.length` or by 

>         // assigning to the `slot` member inside inline assembly.

Author: It might be better to break the lines earlier. So that we don't need to scroll horizontally https://5f7d8f1c9cb1280007c7a477--solidity-blog.netlify.app/2020/10/06/solidity-dynamic-array-cleanup-bug/

You are not affected if you are only using `.push(<arg>)` or if you assign a value (even zero) to
the new elements after increasing the length of the array.

## How to check if contract is vulnerable

If your contract does all of the following:
  1. contract has a dynamically-sized array in storage whose base-type is of size at most 16 bytes,
  2. contract performed an assignment to the aforementioned array by another array of smaller

>      length,

Reviewer: ```suggestion
     length (from either memory, storage or calldata),
```

>   3. if you resize the aforementioned array without assigning a value to the resized indices,

Reviewer: ```suggestion
  3. you increase the length of the aforementioned array without assigning a value to the new elements,
```

  2. Contract performed an assignment to the aforementioned array by another array of smaller
     length (from either memory, storage or calldata).

>   3. You increase the length of the aforementioned array without assigning a value to the new elements.

Author: ```suggestion
  3. You increased the length of the aforementioned array without assigning a value to the new elements.
```

---
layout: post
published: true
title:  "Solidity Dynamic Array Cleanup Bug"
date:   2020-10-06
author: Solidity Team
category: Security Alerts
---

> On September 17, 2020, a bug in the Solidity code generator was found. The bug is fixed with version 0.7.3

Reviewer 3: ```suggestion
On September 17, 2020, a bug in the Solidity code generator was found. The bug is fixed with [version 0.7.3](https://github.com/ethereum/solidity/releases/tag/v0.7.3)
```

> released on October 6, 2020. The bug is present in all prior versions of Solidity.

Reviewer 3: ```suggestion
released on October 7, 2020. **The bug is present in all prior versions of Solidity.**
```

> We assigned the bug, a severity level of *medium*.

Reviewer 3: ```suggestion
We assigned the bug a severity level of "medium".
```

> ## Dynamic Array Cleanup Bug

Reviewer 3: ```suggestion
## Technical Details of the Bug
```

Reviewer 3: Otherwise the article of the article is "Dynamic Array Cleanup Bug" and the sub-heading is the same.

**Summary**: When assigning a dynamically-sized array with types of size at most 16 bytes in storage

> causing the assigned array to shrink, some parts of deleted slots were not zeroed out.

Reviewer 3: ```suggestion
causing the assigned array to shrink, some parts of the deleted slots were not zeroed out.
```

This sentence is quite hard to read, can we simplify it?

Author: Would something like "For a dynamically-sized storage-array, with types of size at most 16 bytes, assignments that require deleting slots did not zero out the deleted slots properly" be better?

the unclean parts of the slot contained the original value, instead of 0. The resizing here is
performed by assigning to the array `length`, by a `push()` or via inline assembly. 

An example of the bug is the following:

```solidity
// SPDX-License-Identifier: GPL-3.0
// The bug is present in all versions of solidity prior to 0.7.3
pragma solidity >=0.6.0 <0.7.3;
contract C {
    uint128[] x;
    function f() public {
        x.push(42); x.push(42); x.push(42); x.push(42);
        uint128[] memory y = new uint128[](1);
        y[0] = 23;
        // This will shrink the array x to one element.
        x = y;
        // Resizing the array to length 4.
        x.push(); x.push(); x.push();
        // After resizing the array, its contents are [23, 0, 0, 42], instead of
        // [23, 0, 0, 0]. Resizing can be also be done by assigning to `.length` or by 
        // assigning to the `slot` member inside inline assembly.
    }
}
```

You are not affected if you are only using `.push(<arg>)` or if you assign a value (even zero) to
the new elements after increasing the length of the array.

## How to check if contract is vulnerable

> If your contract does all of the following:

Reviewer 3: ```suggestion
If your contract does all of the following, try to run tests to see if your contract was relying on having elements in the resized part to be zero.
```

>   1. contract has a dynamically-sized array in storage whose base-type is of size at most 16 bytes,

Reviewer 3: ```suggestion
  1. Contract has a dynamically-sized array in storage whose base-type is of size at most 16 bytes.
```

>   2. contract performed an assignment to the aforementioned array by another array of smaller

Reviewer 3: ```suggestion
  2. Contract performed an assignment to the aforementioned array by another array of smaller
```

>      length (from either memory, storage or calldata),

Reviewer 3: ```suggestion
     length (from either memory, storage or calldata).
```

>   3. you increase the length of the aforementioned array without assigning a value to the new elements,

Reviewer 3: ```suggestion
  3. You increase the length of the aforementioned array without assigning a value to the new elements.
```

>   4. you read from one of the new elements before assigning a new value.

Reviewer 3: ```suggestion
  4. You read from one of the new elements before assigning a new value.
```

> try to run tests to see if your contract was relying on having elements in the resized part to be zero.

Reviewer 3: ```suggestion
```
